taichi/ir/control_flow_graph.cpp

```diff
@@ -134,20 +139,36 @@ Stmt *CFGNode::get_store_forwarding_data(Stmt *var, int position) const {
       break;
     }
   }
+
+  // Check if store_stmt will ever influence the value of var
   auto may_contain_address = [](Stmt *store_stmt, Stmt *var) {
-    for (auto store_ptr : irpass::analysis::get_store_destination(store_stmt)) {
-      if (irpass::analysis::maybe_same_address(var, store_ptr)) {
-        return true;
+    std::vector<Stmt *> aliased_vars = {var};
+    if (var->is<MatrixPtrStmt>()) {
+      aliased_vars.push_back(var->as<MatrixPtrStmt>()->origin);
+    }
+    for (auto aliased_var : aliased_vars) {
+      for (auto store_ptr : irpass::analysis::get_store_destination(
+               store_stmt, true /*get_aliased*/)) {
+        if (irpass::analysis::maybe_same_address(aliased_var, store_ptr)) {
+          return true;
+        }
       }
     }
     return false;
   };
 
+  // Check for aliased address
+  // There's a store to the same dest_addr before this stmt
   if (last_def_position != -1) {
-    // The UD-chain is inside this node.
+    // result: the value to store
     Stmt *result = irpass::analysis::get_store_data(
         block->statements[last_def_position].get());
-    if (!var->is<AllocaStmt>()) {
+    bool is_tensor_involved = result && (result->ret_type->is<TensorType>() ||
+                                         var->ret_type->is<TensorType>());
+    if (!var->is<AllocaStmt>() || is_tensor_involved) {
```

Comment:
It would be clearer if it is changed to `!(var->is<AllocaStmt>() && !is_tensor_involved)` if you want to skip if the var is a scalar alloca stmt